packages/ymax-planner/src/cosmos-rest-client.ts

```diff
@@ -0,0 +1,252 @@
+  denom: string;
+  amount: string;
+}
+
+interface BalanceResponse {
+  balances: Coin[];
+  pagination?: {
+    next_key: string | null;
+    total: string;
+  };
+}
+
+interface CosmosApiError extends Error {
+  statusCode: number;
+  chainId: string;
+  endpoint?: string;
+}
+
+// Predefined chain configurations
+const CHAIN_CONFIGS: Record<string, Record<string, ChainConfig>> = {
+  main: {
+    noble: {
+      chainId: 'noble-1',
+      restEndpoint: 'https://noble-api.polkachu.com',
+      name: 'Noble',
+    },
+    agoric: {
+      chainId: 'agoric-3',
+      restEndpoint: 'https://main.api.agoric.net',
+      name: 'Agoric',
+    },
+  },
+  devnet: {
+    noble: {
+      chainId: 'grand-1',
+      restEndpoint: 'https://noble-testnet-api.polkachu.com:443',
+      name: 'Noble',
+    },
+    agoric: {
+      chainId: 'agoricdev-25',
+      restEndpoint: 'https://devnet.api.agoric.net',
+      name: 'Agoric',
+    },
+  },
+};
+
+export class CosmosRestClient {
+  private readonly timeout: number;
+
+  private readonly retries: number;
+
+  private readonly variant: string;
+
+  private readonly chainConfigs: Map<string, ChainConfig>;
+
+  constructor(config: CosmosRestClientConfig = {}) {
+    this.timeout = config.timeout ?? 10000; // 10s timeout
+    this.retries = config.retries ?? 3;
+    this.variant = config.variant ?? 'devnet';
+
+    const chainConfig = CHAIN_CONFIGS[this.variant];
+
+    if (!chainConfig) {
+      throw new Error(`Unknown chain config ${this.variant}`);
+    }
+
+    // Initialize with predefined chains
+    this.chainConfigs = new Map(Object.entries(chainConfig));
+  }
+
+  /**
+   * Add or update a chain configuration
+   */
+  addChain(chainKey: string, chainConfig: ChainConfig): void {
+    this.chainConfigs.set(chainKey, chainConfig);
+  }
+
+  /**
+   * Get available chain configurations
+   */
+  getAvailableChains(): Array<{ key: string; config: ChainConfig }> {
+    return Array.from(this.chainConfigs.entries()).map(([key, config]) => ({
+      key,
+      config,
+    }));
+  }
+
+  /**
+   * Fetch account balances for a specific chain
+   */
+  async getAccountBalances(
+    chainKey: string,
+    address: string,
+    pagination?: { limit?: number; offset?: number },
+  ): Promise<BalanceResponse> {
+    const chainConfig = this.chainConfigs.get(chainKey);
+    if (!chainConfig) {
+      throw new Error(
+        `Chain configuration not found for: ${chainKey}. Available: ${Array.from(this.chainConfigs.keys()).join(', ')}`,
+      );
+    }
+
+    let url = `${chainConfig.restEndpoint}/cosmos/bank/v1beta1/balances/${address}`;
+
+    // Add pagination parameters if provided
+    const params = new URLSearchParams();
+    if (pagination?.limit) {
+      params.set('pagination.limit', pagination.limit.toString());
+    }
+    if (pagination?.offset) {
+      params.set('pagination.offset', pagination.offset.toString());
+    }
+    if (params.size > 0) {
+      url += `?${params.toString()}`;
+    }
```

Comment:
Repository conventions:
```suggestion
    // Add pagination parameters if provided
    const params = new URLSearchParams();
    if (pagination?.limit) {
      params.set('pagination.limit', `${pagination.limit}`);
    }
    if (pagination?.offset) {
      params.set('pagination.offset', `${pagination.offset}`);
    }
    if (params.size > 0) {
      url += `?${params.toString()}`;
    }
```